### DISCUSSION TITLE
Flux vs Microsoft Extension

### DISCUSSION THREAD

**Discussion by Thaval** [fluxcd/flux2#2635] [Q&A] [answered]
Hi, 

while finding out how to implement Flux, I humbled upon your documentation and the documentation provided by Microsoft (https://docs.microsoft.com/de-de/azure/azure-arc/kubernetes/tutorial-use-gitops-flux2).
To decide which one I will choose, I need to figure out what are the differences? 

Unfortunately, the documentations are not in depth enough, so not everything is covered up (regarding to questions like "why, how").

Can someone help me out here?

**Comment by kingdonb** [comment, marked as answer]
Here's my understanding so far, rudimentary:

1. `Microsoft.flux` extension can be installed on Azure AKS clusters, since it is embedded in the Azure platform. If you are on AKS, then installing Flux as an addon is the simplest option, (much simpler than Bootstrapping a "Generic" Flux installation) – however, as a significant drawback, by undertaking this process you learned almost nothing.\
\
But if you already know Flux, and you are managing infrastructure with Azure, then sure, simpler is better. Azure can manage updates basically on your behalf. There is no "golden repo" or bootstrap repo which you must own that contains the Flux cluster admin entrypoint to protect.\
\
There's nothing stopping you from creating Flux kustomizations with Cluster scope (well, I suppose policy could block it.) It is nice to have Flux managed for you.
3. Azure Arc can be installed on any K8s cluster, whether it is Azure or not. Bare metal, kind cluster that runs on NAT, doesn't matter. You will get Azure management features on that cluster, as long as the agent can dial the Azure control plane from wherever it sits. Ostensibly you are managing your own control plane, but it gets monitoring, an agent runs on it that maintains and upgrades Azure services, and this includes the GitOps (Microsoft.flux) extension, as I understand it.\
\
So you can have the same "no bootstraps" experience on any cluster, and pay a smaller (per cluster-vCPU) management fee to use Azure control panels on your own cluster.
6. Generic Flux is the original Flux experience.\
\
If you read the docs at https://fluxcd.io/docs and follow all of our guides, and track the latest version of Flux, then you can be guaranteed you get the full Flux experience, subject to no other schedule than your own, only by installing Flux on your actual own clusters by yourself. Microsoft may stay reasonably current with upgrades, or they may batch them up quarterly for you, (I honestly don't know Microsoft's release schedule for Flux upgrades or anything else) ...

But I'm reasonably sure they don't keep pace with Flux's ~weeklyish minor releases and more frequent patches pattern of updating, and in fairness probably most of their customers don't want that property of fast iteration in infrastructure. Fast responsiveness to CVEs and upgrades for security issues is another matter. (I'm sure Microsoft, as a well-known Flux infrastructure builder, is keen to be responsive to security concerns, and I'm sure they are subscribed and reading our CVE bulletins and staying ahead of them as well.)

So if you want to collaborate with Flux developers as they work and receive fixes for any issues you might uncover in real time as they're being fixed, then bootstrapping generic CNCF Flux according to the getting started guide and subsequent guides is potentially your best option since you can easily [manage upgrades](https://fluxcd.io/docs/installation/#bootstrap-upgrade) yourself (and even automate them, as pull requests!) If the Azure control plane feature list is appealing to you, on the other hand, you have AKS and Azure Arc which you can choose from, that are both I'm sure also great options.

This is my rudimentary understanding after trying both Microsoft Flux products for the first time today, (thanks for giving me this great opening to clarify my understanding,) since I work on Flux, I feel like I should know all of this, but it's not exactly easy or transparent to figure it all out without actually trying these products to understand what separates them. I've been an Azure user for a while on-again off-again, but just haven't had the drive or occasion to dive into Flux+Azure before. 👍

Edit: OK, I'm back

I just tried installing Azure Arc on my aarch64 testing workstation on a `kind` cluster, and found that it's currently impossible. But...

https://azure.microsoft.com/en-us/updates/public-preview-arm64based-azure-vms-can-deliver-up-to-50-better-priceperformance/

Based on this announcement, I think we can expect to get arm64 support in Azure Arc soon – it's already supported in CNCF Flux!

So, this is one further advantage of bootstrapping in the Flux-specific method today – it already works on arm64, it has been a supported release target since October 2020. Whereas if you try installing the `connectedk8s` extension on an arm64 machine or cluster today, you're only going to waste your time, as it's not on the supported list as of today: https://aka.ms/ArcK8sSupportedOSArchitecture
